Unit test cleanup

majakusber opened the pull request:
Cleaning, extending and adding new unit tests. Since there is no limit to writing unit tests, I expect this does not cover all that you may have in mind and thus any suggestions for more tests are welcome.

Commit:
Renaming java files

Commit:
Cleaning existing unit tests

Commit:
Breaking the old parse_generic_class into two unit tests

A unit test for inner classes and a unit test for a basic generic class

Commit:
Extending test for signature/descriptor mismatch

Commit:
Extending test for generic functions

Commit:
Extending test for generic class

Commit:
Adding unit test for functions with generics

Commit:
Extending and cleaning test for generic class with generic inner classes

Commit:
Renaming unit test for generic inner classes to bounded generic inner classes

Commit:
Adding unit test for nested generics

majakusber commented:
@thk123 @mgudemann @chrisr-diffblue @NlightNFotis I guess this concerns all of us so please take a look.

majakusber commented on unit/java_bytecode/java_bytecode_parse_generics/RecursiveGeneric.java:
```diff
@@ -1,4 +1,4 @@
-class RecursiveGeneric extends SimpleRecursiveGeneric<RecursiveGeneric>
+class RecursiveGeneric<T extends RecursiveGeneric<T>>
```
This was supposed to be just renaming, I will correct this.

majakusber replied:
This is actually correct, the original file `RecursiveGeneric` got renamed to `RecursiveGeneric_Derived`, and the original `SimpleRecursiveGeneric` got in turn renamed to `RecursiveGeneric`

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/GenericArray.java:
```diff
@@ -1,6 +1,4 @@
 class GenericArray<T>
 {
   public T [] t;
```
Is it also worth adding a test for something like a field of type `SomeClass<T>[]` ?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,54 +1,43 @@
-  public static void processUpperBoundInterfaceGeneric(SimpleGeneric<? extends BasicInterface> x) {
-    assert(x.t.getX() == 4);
+  public static void processUpperBoundInterfaceGeneric(Generic<? extends
+  Interface> x)
```
Looks like your editor has done some auto-formatting here - could you make the line wrapping more readable? Maybe break after the `(` and then indent?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,54 +1,43 @@
-  public static void processUpperBoundClassGeneric(SimpleGeneric<? extends Foo> x) {
-    assert(x.t.getX() == 4);
+  public static void processUpperBoundClassGeneric(Generic<? extends
+  Interface_Implementation> x)
```
Ditto with formatting here.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,54 +1,43 @@
-  public static void processLowerBoundGeneric(SimpleGeneric<? super Foo> x, Foo assign) {
-    x.t = assign;
+  public static void processLowerBoundGeneric(Generic<? super
+  Interface_Implementation> x, Interface_Implementation assign)
```
And same again here with the formatting

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,54 +1,43 @@
   // It is not legal Java to specify both an upper and lower bound
-  // public static void processBoundSuperClassGeneric(SimpleGeneric<? extends Object super Foo> x, Foo assign) {
-  //   x.t = assign;
+  // public static void processBoundSuperClassGeneric(Generic<? extends Object super Interface_Implementation> x, Interface_Implementation assign)
```
These commented out parts are presumably not expected to be supported yet? While we may not have unit tests to support these features yet, is it worth still having this code in the .java/.class files anyway to at least test/prove that we at least don't crash when encountering them? Alternatively, if we *do* currently crash when encountering them, is there a Jira issue covering that crash, and if so, it would be nice if that Jira issue was reference from within these comments, e.g. ` // This currently causes a crash, but should be re-enabled once TG-1234 is fixed` - having comments like that helps make it easier to re-enable useful tests once things get fixed, or to at least understand why something is commented out :-)

thk123 replied:
See comment above - this isn't actually valid Java (it isn't clear to me what has even changed on this line?)

majakusber replied:
As Thomas pointed out, this is not valid Java code but it may be a good idea to leave it here, in case someone would want to extend the test and would ponder about the same example ;) The only thing that changed here really is the renaming and removing the assignment.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+
+SCENARIO(
+  "java_bytecode_parse_bounded_generic_inner_classes",
+  "[core][java_bytecode][java_bytecode_parse_generics]")
```
Given that inner classes seem to throw up a few bugs in several parts of test-generator, is it worth us starting to add an additional tag to unit tests involving inner classes? e.g. `[java_bytecode_inner_classes]` ?

majakusber replied:
Good point. Shouldn't it be `[java_bytecode_parse_generics_inner_classes]` though? Or would you make it for general inner classes too? Also, not all tests that have some inner class are focusing specifically on testing aspects of inner classes, some just use it for convenience. Should all of them be tagged then?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+    }
+  }
+
+  THEN("The generic fields should be annotated with concrete types")
```
Should this `THEN` block be inside a `WHEN` block?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+      REQUIRE_FALSE(is_java_generics_class_type(java_class_type));
+
+      // TODO (tkiley): Extend this unit test when bounds are correctly
+      // parsed.
```
Jira issue covering that? ;-)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+      auto generic_param_iterator =
+        java_generics_class_type.generic_types().cbegin();
+
+      // The first parameter should be called K
```
There seems to be a couple of places where we do things like 'first type parameter should be called K' - could the code for that be factored out into a function in `test_utils/require_type` or somewhere similar?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+
+      REQUIRE(java_generics_class_type.generic_types().size() == 2);
+
+      auto generic_param_iterator =
```
What is the reason for using an iterator here? Since you are explicitly accessing the two elements of the vector, one after the other and manually bumping the iterator (rather than doing an actual loop using the iterator), can't you just directly index the elements? e.g. `java_generics_class_type.generic_types().at(0)` and ` java_generics_class_type.generic_types().at(1)` ?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,231 @@
+        REQUIRE(
+          java_generics_class_type_var(1, java_generics_class_type) ==
+          twoelementinner_name + "::V");
+      }
```
Since the code for the first and second parameters is duplicate apart from the identifier name, it's worth thinking about whether this can be factored our into a helper function - probably the same one I previously mentioned :-)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp:
```diff
@@ -23,18 +23,17 @@ SCENARIO(
+    const class_typet &derived_class_type =
       require_symbol::require_complete_class(derived_symbol);
 
     // TODO(tkiley): Currently we do not support extracting information
```
Jira reference would be nice...

thk123 replied:
I agree - I can't find a relevant ticket, if you could create one @svorenova that would be appreciated.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,340 @@
+
+      const symbolt func_symbol =
+        new_symbol_table.lookup_ref(process_func_name);
+      const code_typet func_code = to_code_type(func_symbol.type);
```
Maybe use `require_code` here instead of `to_code_type` ? Or possibly even add a `require_code_symbol` to `testing_utils/require_symbol.cpp` ?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,340 @@
+          REQUIRE(is_java_generic_type(param_x.type()));
+
+          const java_generic_typet generic_x =
+            to_java_generic_type(param_x.type());
```
For this pattern of `is_java_generic_type`, followed by `to_java_generic_type`, it's probably worth factoring out into a `require_java_generic_type` helper function - https://github.com/diffblue/cbmc/pull/1460/files#diff-6dc35e009e58da05abb1a1765acdd9f7R77 might have something you can steal.

thk123 replied:
These look useful Chris - would you mind cherry picking that commit @svorenova (or alternatively you can pull it into a PR @chrisr-diffblue and then base this on top of it).

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_array_class.cpp:
```diff
@@ -23,42 +23,39 @@ SCENARIO(
-  const symbol_typet &subtype=to_symbol_type(t_component.subtype());
-
+  const pointer_typet &t_component = to_pointer_type(
+    type.get_component("t").type());
```
Can you use the `require_pointer` helper function here?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -16,307 +18,148 @@
 
 SCENARIO(
-  "java_bytecode_parse_generics",
+  "java_bytecode_parse_generic_class_one_param",
```
I wonder whether if we have changed the name of the scenario, whether we should also change the filename? I know it's pretty easy to do an all-project search for the scenario name, but it might make sense if all the unit test files are named to match their scenario name?

majakusber replied:
As you mention below, there are two scenarios in the file, the names of the scenarios are prefixed with the name of the file. I discussed with @thk123 and created a task in JIRA (TG-1287) to re-organise the whole folder with unit tests for generics and that would decide the convention. Do you agree with keeping it as it is for now and resolving it in the mentioned task?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -16,307 +18,148 @@
 #include <util/language.h>
 #include <util/message.h>
 #include <java_bytecode/java_bytecode_language.h>
+#include <iostream>
```
Is this left over debug?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -16,307 +18,148 @@
-          symbol_typet &bound_type=to_symbol_type(sub_type);
-          REQUIRE(bound_type.get_identifier()=="java::java.lang.Object");
-        }
+        const java_generic_parametert &generic_param = generic_types[0];
```
I might be tempted here to drop the `generic_types` variable and just use the compound expression entirely - e.g. `const java_generic_parametert &generic_param = java_generics_class_type.generic_types()[0]`

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -16,307 +18,148 @@
-        new_symbol_table
-          .has_symbol("java::generics$bound_element.f:()Ljava/lang/Number;"));
+SCENARIO(
+  "java_bytecode_parse_generic_class_two_param",
```
Ah, I see you've split the scenario into multiple scenarios. I guess there is a design question to ask around whether we want one scenario per file or not. I'm open either way, but worth considering.

majakusber replied:
See comment above.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+
+      require_type::require_pointer(
+        field_component.type(), symbol_typet
+          (class_prefix + "$InnerClass"));
```
Nit: Break after the `(` - it's less likely to trick someone into thinking you are passing three parameters to `require_pointer` :-)

thk123 replied:
Did clang-format produce this?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+        const java_generic_parametert &generic_param = generic_variables[0];
+        REQUIRE(
+          generic_param.type_variable() ==
+          symbol_typet(class_prefix + "::T"));
```
Nit: I'm not sure if this is the new clang-format style or not (apologies for the noise if it is) - but indenting this line a bit would make it easier to see that it is the RHS of the `==` operator. If this is the clang-format style, then just ignore me :-)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+            java_generic_inst_parametert(
+              symbol_typet(
+                "java::Interface_Implementation")));
+        }
```
Consider factoring the above two blocks into a helper function that you can call on `generic_variables[0]` and  `generic_variables[1]` ?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+            symbol_typet(class_prefix + "::T"));
+        }
+      }
+    }
```
Ditto with these two code blocks, which are even more identical to each other.

thk123 replied:
Apologies I originally wrote this  :blush: but I agree this would be better

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+
+#include <java_bytecode/java_bytecode_language.h>
+
+#include <iostream>
```
Left over debug?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+          generic_param.type_variable() ==
+          symbol_typet(class_prefix + "::T"));
+      }
+    }
```
Looks like the above two blocks could be factored out too.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -17,59 +17,1171 @@
+
+          THEN("The bounds are set correctly")
+          {
+            // TODO: the bounds are not parsed yet; extend tests when fixed
```
Jira reference?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -17,59 +17,1171 @@
+
+          THEN("The bounds are set correctly")
+          {
+            // TODO: the bounds are not parsed yet; extend tests when fixed
```
Jira reference?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -17,59 +17,1171 @@
+
+          THEN("The bounds are set correctly")
+          {
+            // TODO: the bounds are not parsed yet; extend the tests when fixed
```
Jira reference?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -17,59 +17,1171 @@
+
+      const symbolt func_symbol =
+        new_symbol_table.lookup_ref(process_func_name);
+      // TODO: should point to something else than Object??
```
Is this TODO something that should be extended now, or something that we are waiting to be fixed elsewhere before extending? Might be nice to clarify this comment, and add refs to Jira if appropriate.

majakusber replied:
I think it relates to bounds as well, I added a reference to the JIRA task I created for bounds.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1339 @@
+
+#include <java_bytecode/java_bytecode_language.h>
+
+#include <iostream>
```
Left over debug?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1339 @@
+          REQUIRE(generic_param ==
+                  java_generic_inst_parametert(symbol_typet("java::Generic")));
+        }
+        // TODO: extend tests when nested generics are parsed correctly
```
Jira reference?

thk123 replied:
I think this is [TG-610] (but I agree would be good to add into the TODO)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp:
```diff
@@ -39,18 +39,28 @@ SCENARIO(
     REQUIRE(func_code.parameters().size()==3);
+
+    // TODO: for now, the parameters are not generic because we fall back to
+    // descriptor due to mismatch; enable tests when fixed
```
Jira reference?

thk123 replied:
Don't think the ticket exists, if you could create it @svorenova that would be great.

chrisr-diffblue reviewed (approved):
In general this looks pretty good and very comprehensive! There's a few minor comments through, which I don't think are blockers, but as general comments, it might be nice if:
a) For any TODO comments, try to add references to Jira issues that they depend on, where possible
b) There's a fair amount of code pattern duplication, so it would probably make the unit tests more readable if a fair amount of that was factored out into files in `unit/testing-utils`. 
As I don't consider these 'blocking' I'm going to approve, but will happily re-review if you feel like addressing these.

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,57 +1,43 @@
   // Test a wild card generic bound by a class
-  // public static void processBoundClassGenericDoubleBound(Generic<? extends Interface_Implementation & Interface> x) {
-  //   assert(x.t.getX() == 4);
+  // public static void processBoundClassGenericDoubleBound(Generic<? extends Interface_Implementation & Interface> x)
```
Though Chris's comment might apply here - or at least the comment should be updated

thk123 commented on unit/testing-utils/require_type.h:
```diff
@@ -17,20 +17,26 @@
+  code_typet require_code(const typet &type);
+
+  code_typet::parametert
+  require_parameter(const code_typet &function_type,
```
I think somehow your clang-format has got confused when formatting here, consider using [these instructions](https://github.com/diffblue/doc-diffblue-common/wiki/Using-clang-format#retroactively-formatting-individual-commits) to fix all the commits to use clang-format

thk123 reviewed (commented):
Reviewed up to "Unit test cleanup" will carry on tomorrow

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,1446 @@
+
+      THEN("The pointer should be generic")
+      {
+        REQUIRE(is_java_generic_type(field_component.type()));
```
I'd be tempted to add a `require_type::java_generic_type` that performs this check and the cast on the next line

thk123 reviewed (commented):
Chris's points are sensible (sorry about the ones that apply to code I wrote originally!) , in general some utilities for the generic stuff would probably help shrink these tests:

`require_instantiated_generic_param(type, specific type)`
`require_non_instantiated_generic_type`

And would help make these tests more robust to future changes to how this is stored. Though it looks like Chris has [written a bunch of these](https://github.com/diffblue/cbmc/pull/1460/files#diff-6dc35e009e58da05abb1a1765acdd9f7R77) so you may want to cherry pick that. 

However, since you've mostly just moved stuff feel free to call out of scope and we can revisit. 

I think the main thing to make sure we do is all the TODOs should reference a JIRA issue and those JIRA issues should reference the unit tests they should extend. 

I've reviewed the commits, will now check to see whether it covers everything but looks like a very good and comprehensive set of tests.

thk123 reviewed (changes requested):
I can't seem to find tests I would associate with Generic fields on classes. I've found ones with doubly nested generics and ones with inner classes so perhaps this is sufficient, but for completeness you may wish to add these. 

Only blocking change as far as I can see is to make sure we've collected all the JIRA issues from the TODOs so they don't get lost.

thk123 commented:
Also, if you've not already spotted the compilation error on CI, be sure to fix that!

Commit:
Extending test for generic arrays

Commit:
Cleaning includes, unifying scenario names, adding JIRA references

Commit:
Applying new function for accessing elements of arrays

Commit:
Applying new utility functions for generics

Commit:
Adding unit test for generic fields

Commit:
Adding a utility for checking java generic class

Commit:
Updating utility functions to check generic/non-generic java classes

Commit:
Refactoring unit test utility functions to make them easier to use

Commit:
Extending test for derived generics

Commit:
Disabling part of unit test due to a bug

Commit:
Deleting a utility function for generics

There is a new one in require_type with the same functionality

Commit:
Adding JIRA tickets cont.

majakusber commented:
Ready to be re-reviewed, @chrisr-diffblue and @mgudemann could you please take a look? ( @thk123 is probably unavailable now)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/WildcardGenericFunctions.java:
```diff
@@ -1,54 +1,47 @@
+  // public static void processBoundClassGenericDoubleBound(
+  //   Generic<? extends Interface_Implementation & Interface> x)
+  // {
+  //
```
Is this commented out case illegal Java too? If so, maybe make that explicit in the comment. If it *is* legal Java, but we currently crash on it it would be nice to have a reference to a Jira issue covering it (sorry for sounding like a broken record...)

majakusber replied:
Good point, it is illegal Java too (double-checked).

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_bounded_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,161 @@
+#include <testing-utils/load_java_class.h>
+#include <testing-utils/require_type.h>
+
+#include <memory>
```
I may be missing where it's used, but is `<memory>` actually needed in this unit test?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_derived_generic_class.cpp:
```diff
@@ -7,37 +7,54 @@
+#include <testing-utils/load_java_class.h>
 
-#include <istream>
 #include <memory>
```
Ditto here - is `<memory>` needed?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,297 @@
+            return_type,
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
+              ".lang.Integer"}});
```
Nit: no need to break the line here between 70/71

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,297 @@
+            return_type,
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
+              ".lang.Integer"}});
```
Nit: no need to break the lines 119/120 here

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,297 @@
+            return_type,
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
+              ".lang.Integer"}});
```
Nit: ditto for line break here, sorry :-)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,297 @@
+            return_type,
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
+              ".lang.Integer"}});
```
Nit: line break again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_functions_with_generics.cpp:
```diff
@@ -0,0 +1,297 @@
+            return_type,
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
+              ".lang.Integer"}});
```
Nit: sorry, another unnecessary line break

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_array_class.cpp:
```diff
@@ -7,58 +7,107 @@
+          element,
+          {{require_type::type_parameter_kindt::Inst,
+            "java::java.lang"
+            ".Integer"}});
```
Nit: don't need the line break here

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp:
```diff
@@ -8,315 +8,111 @@
+#include <testing-utils/require_type.h>
 
-#include <istream>
 #include <memory>
```
Is `<memory>` needed here?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+      {
+        require_type::require_java_generic_type(
+          field_component.type(),
+          {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: I'd be tempted to break this line after the `::Var,`, so that it matches the other components in this initializer list. It makes it easier to read at a glance of the layout style of all the elements is matching :-)

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+      {
+        require_type::require_java_generic_type(
+          field_component.type(),
+          {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Same minor nit here about making the layout of all three elements of the initializer list be formatted in the same layout.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+          {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
+           {require_type::type_parameter_kindt::Inst,
+            "java::Interface_Implementation"},
+           {require_type::type_parameter_kindt::Var, class_prefix + "::T"}});
```
Very minor nit: Again, if it's not too much of a pain, then re-formatting this initializer list so that all elements are broken in the same way would be nice.

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        param_type.type(), symbol_typet(class_prefix + "$GenericInnerClass"));
+      require_type::require_java_generic_type(
+        param_type.type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: Again, would be nice to keep the formatting of the initializer list consistent for all elements

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerClass"));
+      require_type::require_java_generic_type(
+        param_type.type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Same small formatting comment

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerGenericClass"));
+      require_type::require_java_generic_type(
+        param_type.type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Same minor nit on layout of the initializer list

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        param_type.type(), symbol_typet(class_prefix + "$TwoParamInnerClass"));
+      require_type::require_java_generic_type(
+        param_type.type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Same nit on the initiializer list formating

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        param_type.type(), symbol_typet(class_prefix + "$TwoParamInnerClass"));
+      require_type::require_java_generic_type(
+        param_type.type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Ditto for this list

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method1";
+    const std::string func_descriptor =
+      ":()"
```
Probably don't need the line break here?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        symbol_typet(class_prefix + "$GenericInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: consistent formatting of the list again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method4";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```
Minor nit: Probably break this string after the 'Classes$' part?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: list formatting again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method5";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```
Minor nit: again, probably break this string after the `Classes$` part. I'm guessing CLion broke these up in an odd way without making it too obvious...

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method6";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```
Minor nit - same string breaking issue

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+          class_prefix + "$GenericInnerClass$DoublyNestedInnerGenericClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: list layout again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method7";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```
Minor nit - string breaking again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+  {
+    const std::string func_name = ".ret_method8";
+    const std::string func_descriptor =
+      ":()LGenericClassWithGenericInner"
```
Minor nit - string breaking again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        symbol_typet(class_prefix + "$TwoParamInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: list formatting consistency

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class_with_generic_inner_classes.cpp:
```diff
@@ -0,0 +1,918 @@
+        symbol_typet(class_prefix + "$TwoParamInnerClass"));
+      require_type::require_java_generic_type(
+        function_call.return_type(),
+        {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: List formatting consistency

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+#include <testing-utils/load_java_class.h>
+#include <testing-utils/require_type.h>
+
+#include <memory>
```
Is `<memory>` needed?

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
```
Minor nit: Don't need the line break here

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java"
```
Or here

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+            field.type(),
+            {{require_type::type_parameter_kindt::Inst,
+              "java::java.lang.Integer"},
+             {require_type::type_parameter_kindt::Var, class_prefix + "::T"}});
```
Minor nit - break after the `,` for consistency

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+            field.type(),
+            {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
+             {require_type::type_parameter_kindt::Inst,
+              "java::java.lang"
```
Minor nit: don't need the line break

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+        {
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_parameter_kindt::Var, class_prefix + "::T"},
```
Very minor nit: list formatting consistency

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_fields.cpp:
```diff
@@ -0,0 +1,224 @@
+        {
+          require_type::require_java_generic_type(
+            field.type(),
+            {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit: break after the `,`

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -15,61 +15,1015 @@
+          THEN("The bounds are set correctly")
+          {
+            // TODO: the bounds are not parsed yet; extend tests when fixed -
+            // issue TG-1286
```
I like that you've already created a THEN block for this future case

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -15,61 +15,1015 @@
+      const code_typet func_code =
+        require_type::require_code(func_symbol.type, 0);
+      // TODO: should point to something else than Object?? - possibly
+      // connecte dto bounds - issue TG-1286
```
Trivial nit: `connected to`

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_wildcard_function.cpp:
```diff
@@ -7,61 +7,64 @@
+      class_prefix + func_name + func_descriptor;
 
     REQUIRE(new_symbol_table.has_symbol(process_func_name));
+    // TODO extend tests when wildacrds are parsed correctly - TG-675
```
Minor nit (unfortunately copy-and-pasted a few times in this file, so perhaps you can do a search-and-replace to fix) - `wildcards`

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          field_component.type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - consistency of list formatting

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          field_component.type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - list formatting again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          param_type.type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - list formatting again

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          param_type.type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - list formatting

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          function_call.return_type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - list formatting

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_nested_generics.cpp:
```diff
@@ -0,0 +1,1140 @@
+      {
+        require_type::require_java_generic_type(
+          function_call.return_type(),
+          {{require_type::type_parameter_kindt::Inst, "java::Generic"},
```
Minor nit - list formatting

chrisr-diffblue commented on unit/java_bytecode/java_bytecode_parse_generics/parse_signature_descriptor_mismatch.cpp:
```diff
@@ -7,74 +7,93 @@
+#include <testing-utils/load_java_class.h>
+#include <testing-utils/require_type.h>
 
 #include <memory>
```
Is `<memory>` needed?

chrisr-diffblue reviewed (approved):
This looks really great! And many thanks for having added in all the Jira references. I've listed a number of minor formatting nits if you still have the energy to address them. One request I might make is could you review the `#includes` in the unit test files, because I'm wondering if all the includes are actually needed and some of them look like possibly left over from previous version of the unit tests before the helper functions were added?
In principal though, I'd be happy to merge this if you could set my mind at rest about that.
@mgudemann or @thk123  When you review, could you pay attention to the changes in `require_type.cpp`, as I had a hand in those changes so I'm probably too close to the code there.

majakusber commented:
Thank you @chrisr-diffblue for a detailed review, all comments are taken care of now (although some of them still show up above, I assume that's Github's inability to sense the change in formatting). And thank you for the awesome functions, they made everything so much easier! @thk123 I also addressed all your comments and added a test for generic fields.

thk123 commented on unit/java_bytecode/java_bytecode_parse_generics/parse_generic_functions.cpp:
```diff
@@ -7,69 +7,1015 @@
+        new_symbol_table.lookup_ref(process_func_name);
+      const code_typet func_code =
+        require_type::require_code(func_symbol.type, 0);
+      // TODO: should point to something else than Object?? - possibly
```
Since `T` doesn't have a bound for `returnSimpleField` I think there is no TODO needed here since it is already correctly marked as generic

majakusber replied:
So the `Object` here acts as a bound (i.e., no bound)?

thk123 reviewed (approved):
Looks great - thanks for the diligence on linking the TODOs with issues

thk123 commented:
Will merge once CI passes - prod me if I forget :+1:
